lib/mongo/background_thread.rb

```diff
@@ -30,6 +30,14 @@ module BackgroundThread
     #
     # @api public for backwards compatibility only
     def run!
+      if @stop_requested && @thread
+        wait_for_stop
+        if @thread.alive?
+          log_warn("Starting a new background thread in #{self}, but the previous background thread is still running")
+          @thread = nil
```

Comment:
It's possible I'm misunderstanding what's going on here -- it looks like we're removing a reference to a thread that's still running, and then starting a new thread. Is that correct, and if so, is there any danger in leaving the previous background thread to run?

Reply:
Yes, it's not great. The alternative is to not start the new thread, but then the old thread may exit eventually, or the old thread may be broken/stuck. Do you have any thoughts on this?

Reply:
If `@stop_requested` is true, that means that `stop!` has been called on this `BackgroundThread` instance, and that `@thread.kill` has been called. How likely is it that this won't actually kill the thread? Are there particular cases where the `.kill` method doesn't work as intended?

Reply:
It seems that on Ruby 2.6 kill succeeds when the thread is both doing CPU and I/O, as follows:

<pre>
irb(main):033:0> start=Time.now; a=Thread.new{ loop { 2**1000000 } }; a.kill; while a.alive?; sleep 0.01; end; p Time.now-start
0.010240296
=> 0.010240296
irb(main):034:0> 
</pre>

<pre>
irb(main):032:0> start=Time.now; a=Thread.new{ open('/mnt/speed/123').read }; a.kill; while a.alive?; sleep 0.01; end; p Time.now-start
0.010207387
=> 0.010207387
</pre>

However, kill only marks the thread for termination as implied by https://ruby-doc.org/core-2.6.3/Thread.html#method-c-kill and https://ruby-doc.org/core-2.6.3/Thread.html#method-c-exit, and does not immediately terminate the thread:

<pre>
irb(main):005:0> a=Thread.new{ open('abc.xyz').read }; a.kill; a.alive?
=> true
</pre>

Reply:
Thanks for the context. One last question -- if the thread was broken/stuck, would there be any way to test for that and force it to exit?

Reply:
I haven't been able to make a thread get stuck yet, hence how to produce this is currently a mystery to me.

As far as forcing it to exit, this is what Thread#kill is meant to do.

Reply:
Okay, then I think this is fine. Thank you for answering all my questions!